test/cpp/end2end/test_service_impl.cc

```diff
@@ -579,35 +591,43 @@ experimental::ServerUnaryReactor* CallbackTestServiceImpl::Echo(
       Finish(Status::OK);
     }
     void LoopUntilCancelled(int loop_delay_us) {
-      if (!ctx_->IsCancelled()) {
-        alarm_.experimental().Set(
-            gpr_time_add(gpr_now(GPR_CLOCK_REALTIME),
-                         gpr_time_from_micros(loop_delay_us, GPR_TIMESPAN)),
-            [this, loop_delay_us](bool ok) {
-              if (!ok) {
-                EXPECT_TRUE(ctx_->IsCancelled());
-              }
-              LoopUntilCancelled(loop_delay_us);
-            });
-      } else {
-        Finish(Status::CANCELLED);
+      {
+        std::lock_guard<std::mutex> l(alarm_mu_);
+        if (!ctx_->IsCancelled()) {
+          // dont_alarm_anymore_ wouldn't be set either since that is only set
+          // in OnCancel
+          EXPECT_FALSE(dont_alarm_anymore_);
+          alarm_.experimental().Set(
+              gpr_time_add(gpr_now(GPR_CLOCK_REALTIME),
+                           gpr_time_from_micros(loop_delay_us, GPR_TIMESPAN)),
+              [this, loop_delay_us](bool ok) {
+                if (!ok) {
+                  EXPECT_TRUE(ctx_->IsCancelled());
+                }
+                LoopUntilCancelled(loop_delay_us);
+              });
+          return;
+        }
       }
+      Finish(Status::CANCELLED);
     }
 
     CallbackTestServiceImpl* const service_;
     experimental::CallbackServerContext* const ctx_;
     const EchoRequest* const req_;
     EchoResponse* const resp_;
-    Alarm alarm_;
-    bool initial_metadata_sent_{false};
-    bool started_{false};
-    bool on_cancel_invoked_{false};
+    std::mutex alarm_mu_;
+    bool dont_alarm_anymore_ /* GUARDED_BY(alarm_mu_) */ = false;
```

Comment:
I'm not understanding why we're tracking this state about alarms. Isn't the concept here just whether or not the RPC was cancelled?

Reply:
Yeah, that's it. I could call it that but I wanted to be more explicit in how we were using it.

Reply:
Recap from our offline discussion:
dont_alarm_anymore_ is an implementation detail. We should use a variable that tracks cancellation rather than the implementation detail of how we're checking for the cancellation status. We could probably use on_cancel_invoked_ here and rename alarm_mu_ to on_cancel_mu_ or something like that.

Reply:
I just decided to use ctx_->IsCancelled() as the source of truth rather than on_cancel_invoked_ . And now, consistent with the discussion above, OnCancel no longer cancels alarms so alarm_mu_ is not needed anymore. There is still a cancel_mu_ to control the new cancellation CV though.
'